### PR TITLE
ARROW-1563: [C++] Implement logical unary and binary kernels for boolean arrays

### PR DESCRIPTION
**Pull request by xhochy**

### CODE REVIEW

File: cpp/src/arrow/compute/compute-test.cc [hunk 1]
```diff
@@ -1285,5 +1287,255 @@ TEST_F(TestHashKernel, ChunkedArrayInvoke) {
...
+  auto ca1 = std::make_shared<ChunkedArray>(ca1_arrs);
+  std::vector<std::shared_ptr<Array>> ca2_arrs = {a2, a2->Slice(1)};
+  auto ca2 = std::make_shared<ChunkedArray>(ca2_arrs);
+  ASSERT_OK(Invert(&this->ctx_, Datum(ca1), &result));
+  ASSERT_EQ(Datum::CHUNKED_ARRAY, result.kind());
+  std::shared_ptr<ChunkedArray> result_ca = result.chunked_array();
+  ASSERT_TRUE(result_ca->Equals(ca2));
+}
+
+TEST_F(TestBooleanKernel, And) {
```
Comment (pitrou) [inline cpp/src/arrow/compute/compute-test.cc]: Is there a way to share most of the code between `TestBooleanKernel::And`, `TestBooleanKernel::Or` and ``TestBooleanKernel::Xor`? It seems they only differ by the kernel being tested and the actual test values.

File: cpp/src/arrow/compute/kernels/boolean.cc [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+namespace compute {
+
+class InvertKernel : public UnaryKernel {
+  Status Call(FunctionContext* ctx, const Datum& input, Datum* out) override {
+    DCHECK_EQ(Datum::ARRAY, input.kind());
+
+    const ArrayData& in_data = *input.array();
+    ArrayData* result;
+
+    if (out->kind() == Datum::NONE) {
```
Comment (pitrou) [inline cpp/src/arrow/compute/kernels/boolean.cc]: At first sight, it doesn't seem obvious what this is about. Can you comment on this?

File: cpp/src/arrow/compute/kernels/boolean.cc [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+  Status Call(FunctionContext* ctx, const Datum& left, const Datum& right,
+              Datum* out) override {
+    DCHECK_EQ(Datum::ARRAY, right.kind());
+    DCHECK_EQ(Datum::ARRAY, left.kind());
+
+    const ArrayData& left_data = *left.array();
+    const ArrayData& right_data = *right.array();
+    ArrayData* result;
+
+    if (out->kind() == Datum::NONE) {
```
Comment (pitrou) [inline cpp/src/arrow/compute/kernels/boolean.cc]: Ditto here.

File: cpp/src/arrow/compute/kernels/boolean.h [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+/// \param[in] context the FunctionContext
+/// \param[in] value datum to invert
+/// \param[out] out resulting datum
+///
+/// \since 0.11.0
+/// \note API not yet finalized
+ARROW_EXPORT
+Status Invert(FunctionContext* context, const Datum& value, Datum* out);
+
+/// \brief Element-wise AND of two boolean dates
```
Comment (pitrou) [inline cpp/src/arrow/compute/kernels/boolean.h]: "dates" is the plural of "datum"?

File: cpp/src/arrow/util/bit-util-test.cc [hunk 5]
```diff
@@ -395,6 +395,51 @@ TEST(BitmapAnd, Unaligned) {
   }
 }
 
+TEST(BitmapOr, Aligned) {
```
Comment (pitrou) [inline cpp/src/arrow/util/bit-util-test.cc]: Similar to the above, it seems `BitmapAnd`, `BitmapOr` and `BitmapXor` are using exactly the same test harness, except with different result values.

File: cpp/src/arrow/util/bit-util.cc [hunk 6]
```diff
@@ -208,6 +229,70 @@ void UnalignedBitmapAnd(const uint8_t* left, int64_t left_offset, const uint8_t*
   writer.Finish();
 }
 
+void AlignedBitmapOr(const uint8_t* left, int64_t left_offset, const uint8_t* right,
```
Comment (pitrou) [inline cpp/src/arrow/util/bit-util.cc]: I also wonder whether we can factor out the common harness with `BitmapAnd`'s variations (perhaps by templatizing with one or two callables?).

**Review by pitrou — COMMENTED** [review]
Looks fine overall. I think this would be better with less repetition in the code :-)

File: cpp/src/arrow/compute/compute-test.cc (hunk 1 above)
Comment (xhochy) [reply cpp/src/arrow/compute/compute-test.cc]: Yes, done.

File: cpp/src/arrow/compute/kernels/boolean.h (hunk 4 above)
Comment (xhochy) [reply cpp/src/arrow/compute/kernels/boolean.h]: As far as I know yes, but I'm not native speaker.

**Comment by xhochy** [comment]
@pitrou Did some template magic and now there is a lot less repetition.

File: cpp/src/arrow/compute/kernels/boolean.h (hunk 4 above)
Comment (pitrou) [reply cpp/src/arrow/compute/kernels/boolean.h]: Got to ask @wesm on this :-)

File: cpp/src/arrow/compute/compute-test.cc (hunk 1 above)
Comment (pitrou) [reply cpp/src/arrow/compute/compute-test.cc]: Much better! :-D

**Review by pitrou — APPROVED** [review]
+1, LGTM. Thank you!

**Review by wesm — COMMENTED** [review]
Nice, I'm glad to see these implemented. I left some comments; there are some follow up JIRAs that need to be created. I will submit a small patch to improve one aspect of the test suite

File: cpp/src/arrow/compute/kernels/boolean.cc [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+    if (in_data.offset != 0) {
+      RETURN_NOT_OK(CopyBitmap(ctx->memory_pool(), validity_bitmap->data(),
+                               in_data.offset, in_data.length, &validity_bitmap));
+    }
+    result->buffers.push_back(validity_bitmap);
+
+    // Allocate output data buffer
+    std::shared_ptr<Buffer> data_buffer;
+    RETURN_NOT_OK(InvertBitmap(ctx->memory_pool(), in_data.buffers[1]->data(),
+                               in_data.offset, in_data.length, &data_buffer));
```
[the remark this answers is not in the document]
Comment (wesm) [reply cpp/src/arrow/compute/kernels/boolean.cc]: https://issues.apache.org/jira/browse/ARROW-3133

File: cpp/src/arrow/compute/kernels/util-internal.cc [hunk 8]
```diff
@@ -50,6 +51,85 @@ Status InvokeUnaryArrayKernel(FunctionContext* ctx, UnaryKernel* kernel,
...
+    } else {
+      left_start_idx += common_length;
+    }
+
+    if (right_start_idx + common_length == right_array->length()) {
+      right_chunk_idx++;
+      right_start_idx = 0;
+    } else {
+      right_start_idx += common_length;
+    }
```
[the remark this answers is not in the document]
Comment (wesm) [reply cpp/src/arrow/compute/kernels/util-internal.cc]: https://issues.apache.org/jira/browse/ARROW-3134

File: cpp/src/arrow/compute/kernels/boolean.cc [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+  Status Call(FunctionContext* ctx, const Datum& input, Datum* out) override {
+    DCHECK_EQ(Datum::ARRAY, input.kind());
+
+    const ArrayData& in_data = *input.array();
+    ArrayData* result;
+
+    out->value = ArrayData::Make(boolean(), in_data.length);
+    result = out->array().get();
+
+    // Allocate or copy bitmap
```
[the remark this answers is not in the document]
Comment (wesm) [reply cpp/src/arrow/compute/kernels/boolean.cc]: https://issues.apache.org/jira/browse/ARROW-3135

**Pull request closed without merging**